src/projections/s2.cpp

```diff
@@ -0,0 +1,386 @@
+ * The S2 projection functions are adapted from the above link and the S2
+ * Math util functions are adapted from
+ * https://github.com/google/s2geometry/blob/0c4c460bdfe696da303641771f9def900b3e440f/src/s2/util/math/vector.h
+ ****************************************************************************/
+
+#define PJ_LIB__
+
+#include <errno.h>
+#include <math.h>
+
+#include "proj.h"
+#include "proj_internal.h"
+
+/* The six cube faces. */
+namespace { // anonymous namespace
+enum Face {
+    FACE_FRONT  = 0,
+    FACE_RIGHT  = 1,
+    FACE_TOP    = 2,
+    FACE_BACK   = 3,
+    FACE_LEFT   = 4,
+    FACE_BOTTOM = 5
+};
+} // anonymous namespace
+
+enum S2ProjectionType {Linear, Quadratic, Tangent, NoUVtoST};
+std::map<std::string, S2ProjectionType> stringToS2ProjectionType { {"linear", Linear}, {"quadratic", Quadratic}, {"tangent", Tangent}, {"none", NoUVtoST} };
+
+namespace { // anonymous namespace
+struct pj_opaque {
+    enum Face face;
+    double a_squared;
+    double one_minus_f;
+    double one_minus_f_squared;
+    S2ProjectionType UVtoST;
+};
+} // anonymous namespace
+PROJ_HEAD(s2, "S2") "\n\tAzi, Sph";
+
+#define EPS10 1.e-10
+
+/* The four areas on a cube face. AREA_0 is the area of definition,
+ * the other three areas are counted counterclockwise. */
+namespace { // anonymous namespace
+enum Area {
+    AREA_0 = 0,
+    AREA_1 = 1,
+    AREA_2 = 2,
+    AREA_3 = 3
+};
+} // anonymous namespace
+
+// =================================================
+//
+//                  S2 Math Util
+//
+// =================================================
+PJ_XYZ Abs(const PJ_XYZ& p) {
+    return {abs(p.x), abs(p.y), abs(p.z)};
+}
+// return the index of the largest component (fabs)
+int LargestAbsComponent(const PJ_XYZ& p) {
+    PJ_XYZ temp = Abs(p);
+    return temp.x > temp.y ?
+             temp.x > temp.z ? 0 : 2 :
+             temp.y > temp.z ? 1 : 2;
+}
+
+// =================================================
+// 
+//              S2 Projection Functions
+//
+// =================================================
+double STtoUV(double s, S2ProjectionType s2_projection) {
+    switch(s2_projection) {
+        case Linear:
+            return 2 * s - 1;
+            break;
+        case Quadratic:
+            if (s >= 0.5) return (1/3.) * (4*s*s - 1);
+            else return (1/3.) * (1 - 4*(1-s)*(1-s));
+            break;
+        case Tangent:
+            s = std::tan(M_PI_2 * s - M_PI_4);
+            return s + (1.0 / (long{1} << 53)) * s;
+            break;
+        default:
+            return s;
+    }
+}
+
+double UVtoST(double u, S2ProjectionType s2_projection) {
+    switch(s2_projection) {
+        case Linear:
+            return 0.5 * (u + 1);
+            break;
+        case Quadratic:
+            if (u >= 0) return 0.5 * std::sqrt(1 + 3*u);
+            else        return 1 - 0.5 * std::sqrt(1 - 3*u);
+            break;
+        case Tangent:
+        {
+            volatile double a = std::atan(u);
```

Comment:
why volatile ?

Reply:
The purpose of volatile here is to force the truncation to double precision to be consistent.